packages/tasit-action/src/TasitAction.test.js

```diff
@@ -84,109 +77,292 @@ describe("Contract", () => {
+    });
+
+    it("should call a write contract method (send tx) - late subscription", async () => {
+      var rand = Math.floor(Math.random() * Math.floor(1000)).toString();
+      subscription = simpleStorage.setValue(rand);
+      const fakeFn = sinon.fake();
+
+      await mineBlocks(simpleStorage.getProvider(), 15);
+
+      const onMessage = async message => {
+        const { data } = message;
+        const { confirmations } = data;
+
+        if (confirmations >= 7) {
+          fakeFn();
+
+          subscription.removeAllListeners();
+
+          const value = await simpleStorage.getValue();
+          expect(value).to.equal(rand);
+        }
+      };
+
+      await subscription.on("confirmation", onMessage);
+
+      await mineBlocks(simpleStorage.getProvider(), 20);
+
+      expect(fakeFn.called).to.be.true;
+    });
+
+    it("should remove listener after timeout", async () => {
+      subscription = simpleStorage.setValue("hello world");
+      const confirmationFn = sinon.fake();
+      const errorFn = sinon.fake();
+
+      const foreverCallback = async message => {
+        confirmationFn();
+      };
+
+      await subscription.on("confirmation", foreverCallback);
+
+      await subscription.on("error", err => {
+        errorFn();
+      });
+
+      await mineBlocks(simpleStorage.getProvider(), 20);
+
+      expect(confirmationFn.called).to.be.true;
+
+      await new Promise(resolve => setTimeout(resolve, 3000));
+
+      expect(
+        subscription.listenerCount("confirmation"),
+        "listener should be removed after timeout"
+      ).to.be.equals(0);
+
+      expect(errorFn.called).to.be.true;
+    });
+
+    it("should emit error event when orphan/uncle block occurs", async () => {
+      subscription = simpleStorage.setValue("hello world");
 
-        const value = await simpleStorage.getValue();
-        expect(value).to.equal(rand);
+      const confirmationFn = sinon.fake();
+      const errorFn = sinon.fake();
 
-        // UnhandledPromiseRejectionWarning
+      const callback = message => {
+        confirmationFn();
+      };
+
+      await subscription.on("confirmation", callback);
+
+      await subscription.on("error", error => {
+        errorFn();
+        subscription.removeAllListeners();
+
+        // FIXME: Assertion not working
         //expect(1).to.equal(2);
-      }
-    };
+        expect(error.message).to.match(/uncle/);
+      });
+
+      const snapshotId = await createSnapshot(simpleStorage.getProvider());
+
+      await mineBlocks(simpleStorage.getProvider(), 15);
 
-    subscription.on("confirmation", onMessage);
+      expect(confirmationFn.called).to.be.true;
 
-    await mineBlocks(simpleStorage.getProvider(), 8);
+      await revertFromSnapshot(simpleStorage.getProvider(), snapshotId);
+
+      await mineBlocks(simpleStorage.getProvider(), 20);
+
+      // Broadcast same message again (simulate reorg)
+      subscription = simpleStorage.setValue("hello world");
```

Comment:
For it to be a proper uncle situation should we also broadcast some unrelated transactions? Does this test require the nonce to be the same for the test to pass?

Reply:
I'm broadcasting again because the snapshot is taken before transaction been included in a block. Without that line, the test will be waiting forever for confirmation.

Note: ganache snapshot doesn't store mempool transactions, only blocks.

Reply:
Moved to: https://github.com/tasitlabs/TasitSDK/issues/104

Reply:
Great. I'll resolve this one, then.